### PR TITLE
3623 discovery improvements

### PR DESCRIPTION
**Pull request by alvacouch**
1. Don't mangle names of organzations, e.g., "CZO Luquillo" stays as it is when used as an owner. 
2. Arrange for solr_update to refresh all records, not just non-existent ones.
3. Categorize the contained content type of composite resources containing files that correspond to common MIME file types -- and are otherwise undocumented by aggregations -- as higher-level characterizations `Document`, `Spreadsheet`, or `Presentation`, respectively. 

### Pull Request Checklist: 
- [x] Positive Test Case Written by Dev

- [x] Automated Testing

- [x] Sufficient User and Developer Documentation

- [x] Passing Jenkins Build

- [ ] Peer Code review and approval

### Positive Test Case
1. In names of users, organizations -- e.g., "CZO Luquillo" -- are not mangled by discover page after re-indexing. Owners show up as "First Last" rather than "Last, First".
2. Create a resource containing a `.csv`. Discovery by content type lists it as containing a`Spreadsheet`.
3, Create a resource with a `.doc` or `.pdf` file. Discover by content type lists it as containing a `Document`. 
4. Create a resource containing a `.ppt` or `.pptx` file. Discovery by content type lists it as containing a `Presentation`. 
 5. Refreshing the index with management command `solr_update --force` updates all records, in case any are stale.

### CODE REVIEW

File: hs_core/search_indexes.py [hunk 1]
```diff
@@ -100,6 +100,56 @@ def normalize_name(name):
     return normalized.strip()
 
 
+# What extensions indicate each kind of high-level content type
+extension_content_types = {
```
Comment (sblack-usu) [inline hs_core/search_indexes.py]: I wonder if this should be put in config so we can add/update/delete without code changes

File: hs_core/search_indexes.py [hunk 2]
```diff
@@ -100,6 +100,56 @@ def normalize_name(name):
...
+
+    types = set([res.discovery_content_type])  # accumulate high-level content types.
+    exts = set()  # track individual file extensions
+
+    # categorize logical files by type, and files without a logical file by extension.
+    for f in resource.files.all():
+        if f.has_logical_file:
+            candidate_type = type(f.logical_file).get_discovery_content_type()
+            types.add(candidate_type)
+        else:  # collect extensions of files not corresponding to logical metadata
```
Comment (sblack-usu) [inline hs_core/search_indexes.py]: Is this really an either/or situation?  Why not categorize content type by both logical_file as well as extensions?

File: hs_core/search_indexes.py (hunk 1 above)
Comment (alvacouch) [reply hs_core/search_indexes.py]: Which one? local_settings.py or settings.py? Or something else? Willing to move it, but you need to be more specific.....

Comment (sblack-usu) [reply hs_core/search_indexes.py]: It belongs in `settings.py` for the default settings.  `local_settings.py` is not in source control and can be used to override the default settings as needed.

File: hs_core/search_indexes.py (hunk 2 above)
Comment (alvacouch) [reply hs_core/search_indexes.py]: I actually do. Here is the logic: 
 
* Anything with an aggregation is already completely documented beyond doubt. An aggregation is a heavyweight piece of metadata. 
* It is presumed that anything with an aggregation is characterized completely by the aggregation. This is hard evidence of what the resource entails. 
* By contrast, file extensions are a soft determination. People can name something .nc and not mean it! 
* So, I prefer to utilize the hard determination as definitive, and only use extensions when I must. 
* Thus the algorithm marks all files documented by aggregations as accounted for, and then only marks the files that are left over by extension. 
* Alas, there are some files that should have aggregations and don't. These include: 

a. files in generic resources. 
b. files whose aggregations were deleted by a user for whatever reason. 
c. files arising from a time when aggregations were optional. 

etc. 

I have every belief that eventually, all of these will have their own aggregations and this mechanism will be redundant. Thus I have designed in its gradual retirement as aggregations are created.

File: hs_core/search_indexes.py (hunk 1 above)
Comment (alvacouch) [reply hs_core/search_indexes.py]: Done.

**Comment by alvacouch** [comment]
@dtarb @sblack-usu 
Discussed on this Wed's call. Discussion included: 

1. Suggested that I disambiguate between categories `"Generic"` and `"Generic Data"`. I changed `"Generic"` to `"Generic Resource"` to disambiguate. Current implementation is live on https://cuahsi-dev-1.hydroshare.org for your perusal. 

2. Suggested that the inferred types might need to be distinguished from the declared types. E.g., in the case of .nc (`"Multi-dimensional (NetCDF)"`) files. The argument for this would be that the extension inference is less reliable than the .nc aggregation, which checks the .nc file is a valid NetCDF file. I could, e.g., label the inferred files as `"Multi-dimensional (.nc)"` to distinguish them. What is your opinion on this? 

@goodalljl I have one other serious question that did not arise from the discussion. Currently I only categorize extensions in `Generic` and `Composite` resources. Model resources are derived from generic resources and are similarly free-form. Should I be using the same algorithm on them? (There's no reason to do this for single-type resources, as extra files would imply a consistency error!) This would, e.g., report model instances that contain documents.

**Comment by dtarb** [comment]
@alvacouch I do not think we should do anything about unindexed shapefiles.  Users can fix them if they need to.  We have a general principle to not change user data, such as aggregating shapefiles that were not previously aggregated.

**Pull request merged**